### PR TITLE
New PirateSpeak Spice

### PR DESCRIPTION
**Pull request by thejasbabu**
This Spice will help one translate texts into pirate language and helps to have fun.
Given the query
"Pirate speak Hello good morning, my old friend" 
will give "Ahoy jolly good morning, me barnacle-covered shipmate"

https://duck.co/ia/view/piratetalk

### CODE REVIEW

**Comment by daxtheduck** [comment]
@thejasbabu Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by thejasbabu** [comment]
https://duck.co/ia/view/talk_like_a_pirate

**Comment by thejasbabu** [comment]
@moollaza Hii, Can you please Review this. Thank you :)

File: lib/DDG/Spice/Piratetalk.pm [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package DDG::Spice::Piratetalk;
+
+# ABSTRACT: Write an abstract here
+
+# Start at http://docs.duckduckhack.com/walkthroughs/forum-lookup.html if you are new
+# to instant answer development
```
Comment (GuiltyDolphin) [inline lib/DDG/Spice/Piratetalk.pm]: Go through and remove any auto-generated comments.

File: share/spice/piratetalk/piratetalk.css [hunk 2]
```diff
@@ -0,0 +1,55 @@
+.brainy__quote--container {
+    overflow: hidden;
+    margin-left: -2.8em;
+}
+
+.brainy__quote {
```
Comment (GuiltyDolphin) [inline share/spice/piratetalk/piratetalk.css]: I'm guessing you just copy-pasted this? We need to update all of this to reflect new Spice.

File: lib/DDG/Spice/Piratetalk.pm [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+use DDG::Spice;
+
+# Caching - http://docs.duckduckhack.com/backend-reference/api-reference.html#caching
+spice is_cached => 1;
+spice proxy_cache_valid => "200 1d"; # defaults to this automatically
+
+spice wrap_jsonp_callback => 1; # only enable for non-JSONP APIs (i.e. no &callback= parameter)
+
+# API endpoint - http://docs.duckduckhack.com/walkthroughs/forum-lookup.html#api-endpoint
+spice to => 'http://pirate-speak.herokuapp.com/piratespeak?text=$1';
```
Comment (GuiltyDolphin) [inline lib/DDG/Spice/Piratetalk.pm]: Use HTTP Secure.

File: share/spice/piratetalk/piratetalk.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        if (!api_result || api_result.translation==="") {
+            return Spice.failed('piratetalk');
+        }
+
+        // Render the response
+        Spice.add({
+            id: "piratetalk",
+
+            // Customize these properties
+            name: "Talk like a Pirate!",
```
Comment (GuiltyDolphin) [inline share/spice/piratetalk/piratetalk.js]: Get rid of ID and name.

File: t/Piratetalk.t [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    'pirate speak hello' => test_spice(
+        '/js/spice/piratetalk/hello',
+        call_type => 'include',
+        caller => 'DDG::Spice::Piratetalk'
+    ),
+    'pirate talk hello how are you' => test_spice(
+        '/js/spice/piratetalk/hello%20how%20are%20you',
+        call_type => 'include',
+        caller => 'DDG::Spice::Piratetalk'
+    ),
```
Comment (GuiltyDolphin) [inline t/Piratetalk.t]: `translate to pirate speak`, `how do I say X in pirate speak` etc., would all be good triggers.

File: lib/DDG/Spice/Piratetalk.pm (hunk 3 above)
Comment (moollaza) [inline lib/DDG/Spice/Piratetalk.pm]: It's actually preferred to use HTTP if it doesn't redirect to HTTPS. The client is connected to use via HTTPS so they're fine.

File: share/spice/piratetalk/piratetalk.js (hunk 4 above)
Comment (moollaza) [inline share/spice/piratetalk/piratetalk.js]: This is a Spice, so those are still required :)

Only Goodies don't need them for now.

Comment (moollaza) [inline share/spice/piratetalk/piratetalk.js]: The `name` here should be `Answer` though

**Comment by thejasbabu** [comment]
@GuiltyDolphin Thanks for the review, made the updates :) If there's still more left, please let me know.

File: share/spice/piratetalk/piratetalk.js (hunk 4 above)
Comment (GuiltyDolphin) [inline share/spice/piratetalk/piratetalk.js]: @moollaza Ah, alrighty.

Comment (thejasbabu) [inline share/spice/piratetalk/piratetalk.js]: Name field is the one which gets displayed as the header in instant answer tab.

Comment (moollaza) [inline share/spice/piratetalk/piratetalk.js]: > Name field is the one which gets displayed as the header in instant answer tab.

Yes, the `name` should be one of our predefined topics, or `Answer` when it's appropriate. In this case `Anwer` is the best tab name.

**Comment by zekiel** [comment]
@thejasbabu  this is really cool! I wanted to make a yodaspeak IA last year but got sidetracked. http://www.programmableweb.com/api/yodaspeak

**Comment by thejasbabu** [comment]
@zekiel Thanks Zekiel. I'm planning to do that as well once I'm done with the Pirate Talk

**Comment by thejasbabu** [comment]
@moollaza Made changes to the name field to show Answer

**Comment by thejasbabu** [comment]
@moollaza  @GuiltyDolphin Is everything fine? Can we push this to beta? Or is there any more changes to be made?

File: lib/DDG/Spice/Piratetalk.pm [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package DDG::Spice::Piratetalk;
+
+# ABSTRACT: Piratetalk will translate texts into Pirate language when triggered with the trigger words mentioned below.
```
Comment (GuiltyDolphin) [inline lib/DDG/Spice/Piratetalk.pm]: We could do with a *shorter abstract* :wink:

File: share/spice/piratetalk/piratetalk.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+            name: "Answer",
+            data: {
+                title:api_result.translation,
+            },
+            meta: {
+                sourceName: "Pirate",
+                sourceUrl: 'https://pirate-speak.herokuapp.com/',
+            },
+            templates: {     
+                group: 'base',
```
Comment (GuiltyDolphin) [inline share/spice/piratetalk/piratetalk.js]: Why are we using the `base` template? Why not just use `text`?

Comment (thejasbabu) [inline share/spice/piratetalk/piratetalk.js]: @GuiltyDolphin I basically wanted the result to look like quotes, similar to quotes spices. And the result can be a line of words sometime, didnt want to use text template.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=Pirate speak How are you old friend?)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/piratetalk_mobile.png)

File: lib/DDG/Spice/Piratetalk.pm [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+# ABSTRACT: Piratetalk will translate texts into Pirate language when triggered with the trigger words mentioned below. 
+# For example, "Pirate Speak Hello" will return "Ahoy!" 
+
+use DDG::Spice;
+
+spice is_cached => 1;
+spice proxy_cache_valid => "200 1d"; 
+spice wrap_jsonp_callback => 1; 
+
+spice to => 'https://pirate-speak.herokuapp.com/piratespeak?text=$1';
```
Comment (moollaza) [inline lib/DDG/Spice/Piratetalk.pm]: @thejasbabu is this your API or do you know who the maintainer is?

Do you know if this is using a free Heroku plan? If so, I'm worried about the required sleep time for the server, which I imagine will cause the response times to be slow.

File: lib/DDG/Spice/Piratetalk.pm [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+triggers any => 'pirate talk', 'pirate speak', 'translate to pirate';
```
Comment (moollaza) [inline lib/DDG/Spice/Piratetalk.pm]: let's use `startend` instead of `any`

File: share/spice/piratetalk/piratetalk.js (hunk 7 above)
Comment (GuiltyDolphin) [inline share/spice/piratetalk/piratetalk.js]: @thejasbabu In that case we need to modify the CSS to get rid of any quote-marks and 'quote-like' items - we just want the text.

Comment (moollaza) [inline share/spice/piratetalk/piratetalk.js]: For now let's keep things simple and stick to the text template. We only use the base template when needed and in this case the design team has a special design for the quote IA.

Any new IA PRs should use a built-in template unless our design team wants something custom

Comment (GuiltyDolphin) [inline share/spice/piratetalk/piratetalk.js]: @thejasbabu We should be able to simplify things by removing the handlebars and CSS - as @moollaza it's preferable to stick to an existing template (it stresses this in the docs), you can do minor tweaking (change font size, color etc.) in the CSS *without* needing to change template.

**Comment by moollaza** [comment]
@thejasbabu I've updated the IA Page ID and link to match the code and the Perl module.

Typically Perl packages use CamelCase names, so I would expect the Perl package to be `DDG::Spice::PirateTalk` or even `::PirateSpeak` (they're often named after the API/Data Provider).

The `id` should then be `pirate_talk` (or `pirate_speak`)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=Pirate speak How are you old friend?)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/piratetalk_mobile.png)

File: lib/DDG/Spice/Piratetalk.pm (hunk 9 above)
Comment (moollaza) [inline lib/DDG/Spice/Piratetalk.pm]: We're going to need to reign in these triggers so they better indicate intent to translate.

The top auto-complete results for "pirate talk" and "pirate speak" aren't attempting to translate a word/phrase.

triggers like "X to pirate speak", "pirate speak for X", "translate to pirate talk", etc are better.

you can use both `triggers start => '...'` and `triggers end => '...'` and even `triggers startend => '...'` to define several different triggers if you'd like

File: lib/DDG/Spice/Piratetalk.pm (hunk 8 above)
Comment (thejasbabu) [inline lib/DDG/Spice/Piratetalk.pm]: It's hosted in free heroku plan.

**Comment by thejasbabu** [comment]
@GuiltyDolphin Hi, made the appropriate changes, but the issue of heroku still exists. Is it fine if it's in heroku free plan?

File: share/spice/pirate_speak/pirate_speak.css [hunk 10]
```diff
@@ -0,0 +1,14 @@
+.zci__body h3 {
```
Comment (moollaza) [inline share/spice/pirate_speak/pirate_speak.css]: @thejasbabu let's remove this CSS -- We have other IA's that modify/translate and we don't use the quotes there. So far we're only using them when displaying an actual quote.

File: share/spice/pirate_speak/pirate_speak.handlebars [hunk 11]
```diff
@@ -1,2 +1 @@
-
-
+{{title}}
```
Comment (moollaza) [inline share/spice/pirate_speak/pirate_speak.handlebars]: We don't need a Handlebars template 👍 

Seeing as you're using the `text` template, and you've set the `title` we'll render the title for you 👍

File: share/spice/pirate_speak/pirate_speak.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        if (!api_result || api_result.translation ==='') {
+            return Spice.failed('pirate_speak');
+        }
+
+        Spice.add({
+            id: "pirate_speak",
+
+            name: "Answer",
+            data: {
+                title:api_result.translation,
```
Comment (moollaza) [inline share/spice/pirate_speak/pirate_speak.js]: We typically have a `subtitle` that provides context. Let's add a subtitle:

```js
subtitle: "Translate to Pirate Speek: " + $input
```

**Comment by moollaza** [comment]
@thejasbabu I'm closing this PR for now. If you're still interested in finishing, please let me know and I'll gladly re-open it for you!

**Pull request closed without merging**